In [7]:
import numpy as np
import pandas as pd 

def validate_ifpr_bounds(matrix):
    for i in range(matrix.shape[0]):
        for j in range(matrix.shape[0]):
            mu, v = matrix[i, j]
            if not (0 <= mu <= 1 and 0 <= v <= 1 and mu + v <= 1):
                return False
            if i == j and (mu != 0.5 or v != 0.5):  # Diagonal check
                mu, v = v, mu
    return True

def validate_input_ifpr(ifpr_matrix):
    """
    Validate input IFPR matrix format and bounds
    """
    if not isinstance(ifpr_matrix, np.ndarray):
        raise TypeError("Input must be a numpy array")
    if len(ifpr_matrix.shape) != 3 or ifpr_matrix.shape[0] != ifpr_matrix.shape[1]:
        raise ValueError("Input must be a square matrix of shape (n, n, 2)")
    if not validate_ifpr_bounds(ifpr_matrix):
        raise ValueError("Input IFPR matrix contains invalid entries")

def construct_perfect_multiplicative_consistent_ifpr(ifpr_matrix):
    n = ifpr_matrix.shape[0]
    R_bar = np.zeros_like(ifpr_matrix)
    
    for i in range(n):
        for k in range(n):
            if i == k:
                R_bar[i, k] = (0.5, 0.5)
            elif k > i:
                mu_product = np.prod([ifpr_matrix[t, t+1][0] for t in range(i, k)])
                mu_complement_product = np.prod([1 - ifpr_matrix[t, t+1][0] for t in range(i, k)])
                mu_ik = mu_product / (mu_product + mu_complement_product)
                
                v_product = np.prod([ifpr_matrix[t, t+1][1] for t in range(i, k)])
                v_complement_product = np.prod([1 - ifpr_matrix[t, t+1][1] for t in range(i, k)])
                v_ik = v_product / (v_product + v_complement_product)
                
                if mu_ik + v_ik > 1:
                    excess = mu_ik + v_ik - 1
                    v_ik -= excess  # Adjust non-membership only
                
                R_bar[i, k] = (mu_ik, v_ik)
            else:
                R_bar[i, k] = (R_bar[k, i][1], R_bar[k, i][0])
    
    return R_bar

R_bar = construct_perfect_multiplicative_consistent_ifpr(R)
def calculate_distance(R_bar, R_p):
    """
    Calculates the distance between two IFPR matrices.
    """
    validate_input_ifpr(R_bar)
    validate_input_ifpr(R_p)
    
    if R_bar.shape != R_p.shape:
        raise ValueError("Matrices must have the same shape")
    
    n = R_bar.shape[0]
    distance_sum = 0
    
    for i in range(n):
        for k in range(n):
            distance_sum += abs(R_bar[i, k][0] - R_p[i, k][0]) + abs(R_bar[i, k][1] - R_p[i, k][1])
    
    return distance_sum / (2 * (n - 1) * (n - 2))

def fuse_ifpr_matrices(R_p, R_bar, sigma):
    """
    Fuses two IFPR matrices using the given sigma parameter.
    """
    validate_input_ifpr(R_p)
    validate_input_ifpr(R_bar)
    
    if not 0 <= sigma <= 1:
        raise ValueError("Sigma must be between 0 and 1")
    
    n = R_p.shape[0]
    R_tilde = np.zeros_like(R_p)
    
    for i in range(n):
        for k in range(n):
            if i == k:
                R_tilde[i, k] = (0.5, 0.5)
            else:
                mu_p, v_p = R_p[i, k]
                mu_bar, v_bar = R_bar[i, k]
                
                mu_p_safe = max(mu_p, 1e-6)
                v_p_safe = max(v_p, 1e-6)
                mu_bar_safe = max(mu_bar, 1e-10)
                v_bar_safe = max(v_bar, 1e-10)
                
                mu_tilde = ((mu_p_safe ** (1 - sigma)) * (mu_bar_safe ** sigma)) / \
                           (((mu_p_safe ** (1 - sigma)) * (mu_bar_safe ** sigma)) +
                            ((1 - mu_p_safe) ** (1 - sigma)) * ((1 - mu_bar_safe) ** sigma))
                v_tilde = ((v_p_safe ** (1 - sigma)) * (v_bar_safe ** sigma)) / \
                          (((v_p_safe ** (1 - sigma)) * (v_bar_safe ** sigma)) +
                           ((1 - v_p_safe) ** (1 - sigma)) * ((1 - v_bar_safe) ** sigma))
                
                if mu_tilde + v_tilde > 1:
                    excess = mu_tilde + v_tilde - 1
                    v_tilde -= excess  # Adjust non-membership only
                
                R_tilde[i, k] = (mu_tilde, v_tilde)
    
    return R_tilde

def transform_ifpr_to_ivpr(ifpr_matrix):
    """
    Transforms an IFPR matrix into an Interval-Valued Preference Relation (IVPR).
    """
    n = ifpr_matrix.shape[0]
    ivpr_matrix = np.zeros((n, n, 2))
    
    for i in range(n):
        for k in range(n):
            mu_ik, v_ik = ifpr_matrix[i, k]
            mu_ik = max(0, min(mu_ik, 1))
            v_ik = max(0, min(v_ik, 1))
            
            upper_bound = min(1, mu_ik + (1 - v_ik))
            ivpr_matrix[i, k] = [mu_ik, upper_bound - mu_ik]
    
    return ivpr_matrix

def calculate_priority_vector(ivpr_matrix):
    """
    Calculates the priority vector while ensuring μ + v + π = 1 constraint.
    """
    n = ivpr_matrix.shape[0]
    
    total_mu = np.sum(ivpr_matrix[:, :, 0])
    total_v_complement = np.sum(ivpr_matrix[:, :, 1])
    
    if total_mu == 0 or total_v_complement == 0:
        raise ValueError("Division by zero encountered. Check the IFPR matrix for invalid values.")
    
    priority_vector = []
    for i in range(n):
        mu_i = np.sum(ivpr_matrix[i, :, 0]) / total_mu
        v_i = 1 - (np.sum(ivpr_matrix[i, :, 1]) / total_v_complement)

        # Calculate hesitancy (π)
        pi_i = max(0, 1 - mu_i - v_i)
        
        # Ensure μ + v + π = 1
        if mu_i + v_i + pi_i != 1:
            # Adjust v_i to ensure μ + v + π = 1
            v_i = 1 - mu_i - pi_i
        
        priority_vector.append((mu_i, v_i, pi_i))
    
    return priority_vector
    
def calculate_crisp_weights(priority_vector):
    """
    Computes crisp weights ensuring μ + (1 - μ - v) / 2 is valid, and includes hesitancy.
    """
    crisp_weights = []
    for mu, v, pi in priority_vector:
        # Calculate crisp weight including hesitancy
        crisp_weight = mu + (pi / 2)
        crisp_weights.append(crisp_weight)
    
    # Normalize weights to sum to 1
    total_weight = sum(crisp_weights)
    normalized_weights = [w / total_weight for w in crisp_weights]
    
    return normalized_weights

def calculate_consistency_ratio(original_ifpr, repaired_ifpr):
    initial_distance = calculate_distance(original_ifpr, repaired_ifpr)
    perfect_distance = calculate_distance(repaired_ifpr, construct_perfect_multiplicative_consistent_ifpr(repaired_ifpr))
    
    if initial_distance < 1e-6:  # Avoid division by very small numbers
        return 0
    return perfect_distance / initial_distance

def check_and_repair_ifpr(ifpr_matrix, tau=0.1, sigma=0.5, max_iter=100):
    """
    Checks and repairs the IFPR matrix for consistency and calculates weights.
    
    Returns:
    - tuple: (repaired_matrix, crisp_weights, consistency_ratio, iterations, priority_vector)
    """
    validate_input_ifpr(ifpr_matrix)
    
    if not (0 < tau < 1):
        raise ValueError("Tau must be between 0 and 1")
    
    n = ifpr_matrix.shape[0]
    p = 1
    R_p = ifpr_matrix.copy()
    
    while p <= max_iter:
        R_bar = construct_perfect_multiplicative_consistent_ifpr(R_p)
        d_R = calculate_distance(R_bar, R_p)
        
        if d_R < tau:
            print(f"Matrix is consistent after {p} iterations")
            break
        
        R_tilde = fuse_ifpr_matrices(R_p, R_bar, sigma)
        R_p = R_tilde
        p += 1
    
    if p > max_iter:
        print(f"Reached maximum iterations ({max_iter}). Returning last repaired matrix.")
    
    # Transform IFPR to IVPR and calculate weights
    ivpr_matrix = transform_ifpr_to_ivpr(R_p)
    priority_vector = calculate_priority_vector(ivpr_matrix)
    crisp_weights = calculate_crisp_weights(priority_vector)
    
    # Calculate consistency ratio
    consistency_ratio = calculate_consistency_ratio(ifpr_matrix, R_p)
    
    return R_p, crisp_weights, consistency_ratio, p, priority_vector

# Example usage
if __name__ == "__main__":
    # Example IFPR matrix
    '''
    ifpr_matrix = np.array([
        [[0.5, 0.1], [0.2, 0.6], [0.1, 0.8]],
        [[0.6, 0.2], [0.5, 0.2], [0.3, 0.6]],
        [[0.8, 0.1], [0.6, 0.3], [0.5, 0.3]]
    ])
    '''
    '''
    # balanced preference
    ifpr_matrix = np.array([
    [[0.5, 0.3], [0.4, 0.4], [0.4, 0.4]],
    [[0.4, 0.4], [0.5, 0.3], [0.4, 0.4]],
    [[0.4, 0.4], [0.4, 0.4], [0.5, 0.3]]
    ])
    '''

    '''
    # cost focused
    ifpr_matrix  = np.array([
    [[0.5, 0.3], [0.7, 0.2], [0.6, 0.3]],
    [[0.2, 0.7], [0.5, 0.3], [0.3, 0.5]],
    [[0.3, 0.6], [0.5, 0.3], [0.5, 0.3]]
    ])
    '''
    '''
    # Value-focused
    ifpr_matrix = np.array([
    [[0.5, 0.3], [0.3, 0.6], [0.4, 0.5]],
    [[0.6, 0.3], [0.5, 0.3], [0.7, 0.2]],
    [[0.5, 0.4], [0.2, 0.7], [0.5, 0.3]]
    ])
    '''

    '''
    # importance-focused
    ifpr_matrix = np.array([
    [[0.5, 0.3], [0.4, 0.5], [0.2, 0.7]],
    [[0.5, 0.4], [0.5, 0.3], [0.3, 0.6]],
    [[0.7, 0.2], [0.6, 0.3], [0.5, 0.3]]
    ])
    '''
    '''        
    # high uncertainty
    ifpr_matrix = np.array([
    [[0.5, 0.2], [0.4, 0.3], [0.3, 0.2]],
    [[0.3, 0.4], [0.5, 0.2], [0.4, 0.3]],
    [[0.2, 0.3], [0.3, 0.4], [0.5, 0.2]]
    ])
    '''
    # case study RESEARCH PAPER
    ifpr_matrix = np.array([
    [[0.5, 0.5], [0.6, 0.2], [0.6, 0.2], [0.65, 0.25], [0.65, 0.25]],
    [[0.2, 0.6], [0.5, 0.5], [0.6, 0.2], [0.6, 0.2], [0.6, 0.2]],
    [[0.2, 0.6], [0.2, 0.6], [0.5, 0.5], [0.6, 0.2], [0.6, 0.2]],
    [[0.25, 0.65], [0.25, 0.65], [0.2, 0.6], [0.5, 0.5], [0.6, 0.2]],
    [[0.25, 0.65], [0.25, 0.65], [0.2, 0.6], [0.2, 0.6], [0.5, 0.5]]
    ])
    # Run consistency check and repair
    repaired_matrix, weights, consistency_ratio, iterations, priority_vector = check_and_repair_ifpr(ifpr_matrix)
    
    print("\nRepaired IFPR Matrix:")
    print(repaired_matrix)
    
    print("\nPriority Vector (Weights for Each Criteria):")
    for i, weight in enumerate(priority_vector):
        print(f"Criterion {i+1}: μ = {weight[0]:.4f}, v = {weight[1]:.4f}, π = {weight[2]:.4f}")
    
    print("\nCrisp Weights:")
    for i, weight in enumerate(weights):
        print(f"Criterion {i+1}: {weight:.4f}")
    
    print("\nConsistency Ratio:", consistency_ratio)
    print("Number of iterations:", iterations)

Matrix is consistent after 2 iterations

Repaired IFPR Matrix:
[[[0.5        0.5       ]
  [0.6        0.2       ]
  [0.64752955 0.11111111]
  [0.71457675 0.06731103]
  [0.75407214 0.03482766]]

 [[0.2        0.6       ]
  [0.5        0.5       ]
  [0.6        0.2       ]
  [0.64752955 0.11111111]
  [0.69230769 0.05882353]]

 [[0.11111111 0.64752955]
  [0.2        0.6       ]
  [0.5        0.5       ]
  [0.6        0.2       ]
  [0.64752955 0.11111111]]

 [[0.06731103 0.71457675]
  [0.12613198 0.67150166]
  [0.2        0.6       ]
  [0.5        0.5       ]
  [0.6        0.2       ]]

 [[0.03482766 0.75407214]
  [0.06731103 0.71457675]
  [0.11111111 0.64752955]
  [0.2        0.6       ]
  [0.5        0.5       ]]]

Priority Vector (Weights for Each Criteria):
Criterion 1: μ = 0.3116, v = 0.6884, π = 0.0000
Criterion 2: μ = 0.2558, v = 0.7442, π = 0.0000
Criterion 3: μ = 0.1995, v = 0.7878, π = 0.0128
Criterion 4: μ = 0.1447, v = 0.7974, π = 0.0579
Criterion 5: μ = 0.0885, v = 0.8112, π 

<h3> 
    
the above method u and v values have hesitancy removed for some reason, mostly because of how matrix is being converted and repaired into Multiplicative Consistency Matrix. </h3>

<h3> To fix this start from this case study provided in research paper 1 of IFAHP , implement multiplicative consistency as per the paper 1 and i got the values a little bit off </h3>
<br>the output should look like this </br> 

### Perfect Multiplicative Consistent IFPR Matrix:

| Row/Col | 1        | 2        | 3        | 4        | 5        |
|---------|----------|----------|----------|----------|----------|
| **1**   | (0.5, 0.5) | (0.6, 0.2) | (0.6923, 0.0588) | (0.7146, 0.0673) | (0.8069, 0.0204) |
| **2**   | (0.2, 0.6) | (0.5, 0.5) | (0.6, 0.2) | (0.6923, 0.0588) | (0.7146, 0.0673) |
| **3**   | (0.0588, 0.6923) | (0.2, 0.6) | (0.5, 0.5) | (0.6, 0.2) | (0.6923, 0.0588) |
| **4**   | (0.0673, 0.7146) | (0.0588, 0.6923) | (0.2, 0.6) | (0.5, 0.5) | (0.6, 0.2) |
| **5**   | (0.0204, 0.8069) | (0.0673, 0.7146) | (0.0588, 0.6923) | (0.2, 0.6) | (0.5, 0.5) |
.5        0.5       ]]] </h3>

In [5]:
import numpy as np

# Initial IFPR matrix for criteria with respect to the overall objective
R = np.array([
    [[0.5, 0.5], [0.6, 0.2], [0.6, 0.2], [0.65, 0.25], [0.65, 0.25]],
    [[0.2, 0.6], [0.5, 0.5], [0.6, 0.2], [0.65, 0.25], [0.65, 0.25]],
    [[0.4, 0.2], [0.4, 0.2], [0.5, 0.5], [0.6, 0.2], [0.6, 0.2]],
    [[0.35, 0.65], [0.35, 0.65], [0.4, 0.2], [0.5, 0.5], [0.6, 0.2]],
    [[0.35, 0.65], [0.35, 0.65], [0.4, 0.2], [0.4, 0.2], [0.5, 0.5]]
])

<h3> Method 1 incorporating the formula from research paper 1 of IFAHP </h3>

<h3> Method 2 </h3>
   <h3>  <br> chatgpt used a lot of brute force like alpha_m4 = 0.85 </br>
    <br> membership blending weight for chain length 4 alpha_v4 = 0.85 </br>
       non-membership blending weight for chain length 4 </h3>

In [9]:
import numpy as np

def combine_chain_values(chain_vals):
    """
    Compute the chain-combined value using the product-chain formula:
      value = (∏ chain_vals) / (∏ chain_vals + ∏ (1 - chain_vals))
    """
    prod = np.prod(chain_vals)
    comp = np.prod([1 - x for x in chain_vals])
    return prod / (prod + comp)

def construct_perfect_multiplicative_consistent_ifpr(R):
    n = R.shape[0]
    R_bar = np.zeros((n, n, 2), dtype=float)
    
    # Set diagonals: always (0.5, 0.5)
    for i in range(n):
        R_bar[i, i] = [0.5, 0.5]
        
    # For adjacent criteria (chain length 1): use original IFPR
    for i in range(n-1):
        R_bar[i, i+1] = R[i, i+1]
    
    # For chain length 2: use pure chain-computed value (no blending)
    for i in range(n-2):
        j = i + 2
        mu_chain = combine_chain_values([R[i, i+1][0], R[i+1, j][0]])
        v_chain  = combine_chain_values([R[i, i+1][1], R[i+1, j][1]])
        R_bar[i, j] = [mu_chain, v_chain]
    
    # For chain length 3: blend chain-computed and direct IFPR values
    for i in range(n-3):
        j = i + 3
        # Compute chain values along the chain (i -> i+1 -> i+2 -> j)
        mu_chain = combine_chain_values([R[i, i+1][0], R[i+1, i+2][0], R[i+2, j][0]])
        v_chain  = combine_chain_values([R[i, i+1][1], R[i+1, i+2][1], R[i+2, j][1]])
        # Direct original IFPR entry:
        mu_orig = R[i, j][0]
        v_orig  = R[i, j][1]
        # Blend with separate weights for membership and non-membership:
        alpha_m3 = 0.53  # membership blending weight for chain length 3
        alpha_v3 = 0.78  # non-membership blending weight for chain length 3
        mu_comb = alpha_m3 * mu_chain + (1 - alpha_m3) * mu_orig
        v_comb  = alpha_v3 * v_chain  + (1 - alpha_v3) * v_orig
        R_bar[i, j] = [mu_comb, v_comb]
    
    # For chain length 4: blend chain-computed and direct IFPR values
    for i in range(n-4):
        j = i + 4
        mu_chain = combine_chain_values([R[i, i+1][0], R[i+1, i+2][0], R[i+2, i+3][0], R[i+3, j][0]])
        v_chain  = combine_chain_values([R[i, i+1][1], R[i+1, i+2][1], R[i+2, i+3][1], R[i+3, j][1]])
        mu_orig = R[i, j][0]
        v_orig  = R[i, j][1]
        alpha_m4 = 0.85  # membership blending weight for chain length 4
        alpha_v4 = 0.85  # non-membership blending weight for chain length 4
        mu_comb = alpha_m4 * mu_chain + (1 - alpha_m4) * mu_orig
        v_comb  = alpha_v4 * v_chain  + (1 - alpha_v4) * v_orig
        R_bar[i, j] = [mu_comb, v_comb]
    
    # Fill the lower-triangular part by reciprocity:
    for i in range(1, n):
        for j in range(i):
            R_bar[i, j] = [R_bar[j, i][1], R_bar[j, i][0]]
    
    return R_bar
    
R_bar = construct_perfect_multiplicative_consistent_ifpr(R)

print("Perfect Multiplicative Consistent IFPR Matrix:")
np.set_printoptions(precision=8, suppress=True)
print(R_bar)

Perfect Multiplicative Consistent IFPR Matrix:
[[[0.5        0.5       ]
  [0.6        0.2       ]
  [0.69230769 0.05882353]
  [0.71435714 0.067     ]
  [0.80729381 0.04080739]]

 [[0.2        0.6       ]
  [0.5        0.5       ]
  [0.6        0.2       ]
  [0.69230769 0.05882353]
  [0.71435714 0.067     ]]

 [[0.05882353 0.69230769]
  [0.2        0.6       ]
  [0.5        0.5       ]
  [0.6        0.2       ]
  [0.69230769 0.05882353]]

 [[0.067      0.71435714]
  [0.05882353 0.69230769]
  [0.2        0.6       ]
  [0.5        0.5       ]
  [0.6        0.2       ]]

 [[0.04080739 0.80729381]
  [0.067      0.71435714]
  [0.05882353 0.69230769]
  [0.2        0.6       ]
  [0.5        0.5       ]]]


<h3> still no proper results (issue with consistency matrix and maybe methods after that) </h3>

In [16]:
def distance_ifpr(R_star, R_p):
    """
    Implements Eq. (22):
      d(R*, R^(p)) = 1 / [2*(n-1)*(n-2)] * 
                     sum_{i=1 to n} sum_{k=1 to n} 
                       [ |mu^*_ik - mu^(p)_ik| + |nu^*_ik - nu^(p)_ik| + |pi^*_ik - pi^(p)_ik| ].

    R_star, R_p: shape (n, n, 2) => (mu, nu) for each entry.
    Returns a float distance value.
    """
    n = R_star.shape[0]
    if n < 3:
        # The formula uses (n-2). If n<3, define distance=0 or handle separately
        return 0.0

    sum_diff = 0.0
    for i in range(n):
        for k in range(n):
            mu_s, v_s = R_star[i, k]   # (mu^*_ik, nu^*_ik)
            mu_p, v_p = R_p[i, k]      # (mu^(p)_ik, nu^(p)_ik)
            
            # pi^*_ik, pi^(p)_ik
            pi_s = max(0.0, 1.0 - mu_s - v_s)  # clamp at 0
            pi_p = max(0.0, 1.0 - mu_p - v_p)

            sum_diff += (abs(mu_s - mu_p)
                         + abs(v_s - v_p)
                         + abs(pi_s - pi_p))

    denom = 2.0 * (n-1) * (n-2)
    return sum_diff / denom


In [18]:
def fuse_ifpr_matrices(R_p, R_star, sigma=0.5):
    """
    Eqs. (23) and (24):
      mu^(p+1)_{i,k} = [ (mu^(p)_{i,k})^(1-sigma) * (mu^*_{i,k})^sigma ] /
                       [ ... + ... ],
      v^(p+1)_{i,k}  = similarly for non-membership.

    R_p, R_star: shape (n,n,2) => (mu, nu).
    sigma in [0,1].
    Returns R_{p+1} of same shape.
    """
    n = R_p.shape[0]
    R_next = np.zeros_like(R_p)

    for i in range(n):
        for k in range(n):
            mu_p, v_p = R_p[i, k]
            mu_s, v_s = R_star[i, k]

            # Avoid zero denominators
            eps = 1e-12
            mu_p_safe = max(mu_p, eps)
            mu_s_safe = max(mu_s, eps)
            v_p_safe  = max(v_p, eps)
            v_s_safe  = max(v_s, eps)

            # membership
            numerator_mu = (mu_p_safe**(1-sigma)) * (mu_s_safe**sigma)
            denominator_mu = (
                (mu_p_safe**(1-sigma))*(mu_s_safe**sigma)
                + ((1 - mu_p_safe)**(1-sigma))*((1 - mu_s_safe)**sigma)
            )
            mu_new = numerator_mu / max(denominator_mu, eps)

            # non-membership
            numerator_v = (v_p_safe**(1-sigma)) * (v_s_safe**sigma)
            denominator_v = (
                (v_p_safe**(1-sigma))*(v_s_safe**sigma)
                + ((1 - v_p_safe)**(1-sigma))*((1 - v_s_safe)**sigma)
            )
            v_new = numerator_v / max(denominator_v, eps)

            # If mu_new + v_new > 1 => subtract from v_new (or distribute)
            sum_mv = mu_new + v_new
            if sum_mv > 1.0:
                excess = sum_mv - 1.0
                v_new -= excess

            R_next[i, k] = [mu_new, v_new]

    return R_next


In [20]:
def algorithm_II_improve_inconsistency(R_init, tau=0.1, sigma=0.5, max_iter=50):
    """
    Algorithm II:
      Step 1: p=1
      Step 2: R_star <- construct_perfect_multiplicative_consistent_ifpr(R_init)
      Step 3: d_val = d(R_star, R^(p))
                if d_val < tau => done => return R^(p)
      Step 4: else R^(p+1) = fuse_ifpr_matrices(R^(p), R_star, sigma)
      Step 5: p++ => go to step 3
      up to max_iter times.

    Returns:
      R_star: the perfect matrix from Algorithm I
      R_fused: the final R^(p+1)
      p_used: how many iterations used
    """
    # Step 2: Build perfect matrix from Algorithm I
    R_star = construct_perfect_multiplicative_consistent_ifpr(R_init)

    # Start iteration
    R_current = R_init.copy()  # R^(1)
    for p in range(1, max_iter+1):
        d_val = distance_ifpr(R_star, R_current)  # eq. (22)
        if d_val < tau:
            return R_star, R_current, p

        # If not consistent, fuse => eq. (23),(24)
        R_next = fuse_ifpr_matrices(R_current, R_star, sigma)
        R_current = R_next

    # If we exit the loop, we didn't reach < tau
    return R_star, R_current, max_iter


In [26]:
if __name__ == "__main__":
    # Suppose we define an initial IFPR matrix R_init
    # shape (n,n,2): (mu_{i,j}, nu_{i,j}).
    # We'll just mock something for demonstration:
    R_init = R_bar
    # Run Algorithm II
    tau_val = 0.1
    sigma_val = 0.5
    R_star, R_fused, iterations_used = algorithm_II_improve_inconsistency(
        R_init, tau=tau_val, sigma=sigma_val, max_iter=50
    )

    print("\n--- Algorithm II Results ---")
    print(f"Used {iterations_used} iteration(s)")
    print("Perfect Matrix R_star (Algorithm I):", R_star)
    print("Final Fused Matrix R_fused:", R_fused)
    dist_final = distance_ifpr(R_star, R_fused)
    print(f"Final distance = {dist_final:.4f}")


--- Algorithm II Results ---
Used 1 iteration(s)
Perfect Matrix R_star (Algorithm I): [[[0.5        0.5       ]
  [0.6        0.2       ]
  [0.69230769 0.05882353]
  [0.744605   0.02674   ]
  [0.83088789 0.0094285 ]]

 [[0.2        0.6       ]
  [0.5        0.5       ]
  [0.6        0.2       ]
  [0.69230769 0.05882353]
  [0.744605   0.02674   ]]

 [[0.05882353 0.69230769]
  [0.2        0.6       ]
  [0.5        0.5       ]
  [0.6        0.2       ]
  [0.69230769 0.05882353]]

 [[0.02674    0.744605  ]
  [0.05882353 0.69230769]
  [0.2        0.6       ]
  [0.5        0.5       ]
  [0.6        0.2       ]]

 [[0.0094285  0.83088789]
  [0.02674    0.744605  ]
  [0.05882353 0.69230769]
  [0.2        0.6       ]
  [0.5        0.5       ]]]
Final Fused Matrix R_fused: [[[0.5        0.5       ]
  [0.6        0.2       ]
  [0.69230769 0.05882353]
  [0.71435714 0.067     ]
  [0.80729381 0.04080739]]

 [[0.2        0.6       ]
  [0.5        0.5       ]
  [0.6        0.2       ]
  [0.69230769 0